// Copyright 2009 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Protocol buffer used by the pagespeed library rule checkers to
// represent violations, potential savings and recommendations.

syntax = "proto2";

option optimize_for = LITE_RUNTIME;

package pagespeed;

message Savings {
  // DNS requests that could be saved if page was optimized.
  optional int32 dns_requests_saved = 1;

  // Number of request that could be saved if page was optimized.
  optional int32 requests_saved = 2;

  // Download bytes that could be saved if page was optimized.
  optional int32 response_bytes_saved = 3;
}

message InputInformation {
  // Number of resources accessed while rendering the page.
  optional int32 number_resources = 1;

  // Number of hosts among which the page was sharded.
  optional int32 number_hosts = 2;

  // Total server response bytes.
  optional int64 total_response_bytes = 3;
}

// Base class for result details
// Ids 1000 and above are reserved for custom result details.  All
// custom details must have different ids.
message ResultDetails {
  extensions 1000 to max;
}

// Result detail to store image dimension information.
message ImageDimensionDetails {
  extend ResultDetails {
    // In order to attach it to Result.details
    optional ImageDimensionDetails message_set_extension = 4812135;  // 'Img'
  }

  // Natural dimensions of the image element.
  optional int32 expected_height = 1;
  optional int32 expected_width = 2;

  // Actual dimensions of the displayed image element.
  optional int32 actual_height = 3;
  optional int32 actual_width = 4;
}

message Result {
  // Identifier of rule that produced this result.
  required string rule_name = 1;

  // Potential savings if page was optimized.
  optional Savings savings = 2;

  // Resource urls that were processed to compute this result.
  repeated string resource_urls = 3;

  // Full details about the rule violation.
  optional ResultDetails details = 4;

  // Optimized version of the current resource, if applicable.
  optional bytes optimized_content = 5;

  // Total response size before savings.
  optional int32 original_response_bytes = 6;
}

message Results {
  repeated Result results = 1;
}

// Typed format string arguments.
message FormatArgument {
  enum ArgumentType {
    URL = 1;
    STRING_LITERAL = 2;
    INT_LITERAL = 3;
    BYTES = 4;
  }

  required ArgumentType type = 1;
  optional string string_value = 2;
  optional int64 int_value = 3;
}

// Result as text document node that contains the markup type, format
// string, instances and children.
message ResultText {
  // format string.  Use '$<number>' to specify position information.
  optional string format = 1;

  // format string arguments.
  repeated FormatArgument args = 2;

  // Children.
  repeated ResultText children = 3;

  // Optimized version of the current resource, if applicable.
  optional bytes optimized_content = 4;
}
